### ISSUE TITLE
Segmentation fault

### ISSUE THREAD

**Issue by Hakerh400** [zenorogue/hyperrogue#77]
### Issue 1

Steps to reproduce:
1. Go to Camelot
2. Collect Orb of Water
3. Go in random directions
4. Observe that game crashed

See the [screencast](https://www.youtube.com/watch?v=KPtUZSsB6iQ) and [debug log](https://github.com/zenorogue/hyperrogue/files/3345117/debug.txt) (analyzed using [WinDbg](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools)). Use the following [saved game](https://github.com/zenorogue/hyperrogue/files/3345122/hyperrogue.log) to reproduce more easily.

It crashes ~10% of times when Orb of Water is collected in Camelot. What can be seen on the screencast was a lucky attempt when it crashed immediately, but sometimes it requires more than one orb to be collected. Just keep wandering in Camelot and keep collecting orbs of water, it will crash eventually.

### Issue 2

Steps to reproduce:
1. Go near Sulfur Lake
2. Kill an enemy

See the [screencast](https://www.youtube.com/watch?v=s2Wrt25cl7c) (not recorded by us - it's from a twitch stream, but we successfully reproduced it locally). It crashes ~1% of times.

**Comment by Hakerh400** [comment]
Ping @zenorogue

Any update on this? The issue has been open for about a year and still consistently reproduces on master...

**Comment by Quuxplusone** [comment]
(Non-maintainer here.) I just tried to reproduce the "Orb of Water in Camelot" bug on OSX [not Windows] and was unable to reproduce it. You really just wander around in Camelot until you find an Orb of Water sitting in a boat? How long does that take? Do you have a savefile where the Orb is nearby?

I used cheat mode command `'O'` to summon orbs until I got an Orb of Water; collected it; repeated multiple times. However, when you use cheat mode, you don't also get a boat along with the Orb. I suspect that if there's a bug here, it probably relates to trying to exit the Round Table in a boat; but I don't know how to summon a boat inside Camelot to test that hypothesis.

**Comment by Hakerh400** [comment]
> Do you have a savefile where the Orb is nearby?

I literally posted the saved game file. There are also two videos demonstrating the issue, including all required log files and crash info.

**Comment by Quuxplusone** [comment]
> I literally posted the saved game file

Yes, and it doesn't seem to include the Orb of Water that you collected in the video (although I also can't reproduce the intricate dance of knight-killing that you did in the video, so who knows if I'm looking in the right place for the Orb of Water). Do you have a savefile where the Orb is _nearby_, so I could just go pick it up and see the crash happen? Or do you know some way of reliably reproducing the crash?

See 
http://sscce.org/
and
https://gamedev.meta.stackexchange.com/questions/2201/whats-a-minimum-complete-verifiable-example-mcve-and-how-do-i-provide-it

**Comment by Hakerh400** [comment]
Game can be saved only after consuming an orb of safety. And the orb of safety teleports you to the crossroads. The issue is about driving a boat inside the round table in the camelot. Sight range is less than 10 and the distance from the middle of the crossroads to the round table boundary is more than 10 tiles. So, what you are talking about is impossible to achieve (to have a savefile within the round table). Or do you want me to reliably predict the random number generator sequence that is used for generating the world?

**Comment by zenorogue** [comment]
Somehow I missed this...

I could not reproduce any of the two issues. Also I do not see how they could have happened (especially the demon error -- nothing surprising happens here, and at 1% crashes while close to lakes of sulphur, this bug should be reported all the time, as you kill lots of demons in Hell and you are close to lakes of sulphur all the time). I have tried running into these with an automatic tester and the bugs did not happen either.

The debug log is useless without knowing which executable was used. Was it one compiled by yourself? If yes, compiling with debug information (-g) could help.

BTW other than the standard save using Orb of Safety, you can also use the "save" feature in the map editor. However, this does not save large scale structures such as the Camelot, so it is likely that it could not be used to reproduce the Camelot issue.

**Comment by Hakerh400** [comment]
> compiling with debug information (-g) could help.

Ok, I tried to compile from source, but faced a lot of problems. My operating system is Windows 8.1 x64.

1. First, I downloaded [MSYS2](https://www.msys2.org/). To be precise, I downloaded `msys2-x86_64-20190524.exe` from [sourceforge](https://sourceforge.net/projects/msys2/files/Base/x86_64/) ([direct link](https://netix.dl.sourceforge.net/project/msys2/Base/x86_64/msys2-x86_64-20190524.exe)), because [their official repo](http://repo.msys2.org/distrib/x86_64/) website [does not seem to work](https://downforeveryoneorjustme.com/repo.msys2.org) currently. Confirmed file hash integrity.
2. Installed MSYS2 into `C:\ProgramFiles\msys64`
3. Added `C:\ProgramFiles\msys64\usr\bin` and `C:\ProgramFiles\msys64\mingw64\bin` to the PATH system environment variable
4. Ran the following pacman commands (pacman is located in `C:\ProgramFiles\msys64\usr\bin\pacman.exe`) (disable firewall if needed):
4.1. `pacman -S mingw-w64-x86_64-gcc`
4.2. `pacman -S mingw-w64-x86_64-SDL`
4.3. `pacman -S mingw-w64-x86_64-SDL_gfx`
4.4. `pacman -S mingw-w64-x86_64-SDL_mixer`
4.5. `pacman -S mingw-w64-x86_64-glew`
4.6. `pacman -S mingw-w64-x86_64-SDL_ttf` (got the following error, but the package continued installation):
```
/usr/bin/mkdir: missing operand
Try '/usr/bin/mkdir --help' for more information.
/usr/bin/cp: missing file operand
Try '/usr/bin/cp --help' for more information.
/usr/bin/cp: missing file operand
Try '/usr/bin/cp --help' for more information.
/usr/bin/cp: missing file operand
Try '/usr/bin/cp --help' for more information.
error: command (/usr/bin/bash /usr/bin/bash -c . /tmp/alpm_EWE3Zi/.INSTALL; post_install 20180409-1 ) failed to execute correctly
```
5. `pacman` crashed with the following error: `assertion "httpc->drain_total >= data->state.drain" failed: file "http2.c", line 496, function: drain_this
      0 [main] pacman 2008 cygwin_exception::open_stackdumpfile: Dumping stack trace to pacman.exe.stackdump`
6. Deleted `db.lck` and restarted `pacman`, the second time it worked.
7. Cloned `https://github.com/zenorogue/hyperrogue` into a local directory and checked the master branch out
8. Installed `make` into `C:\ProgramData\chocolatey\bin\make.exe` using the [https://chocolatey.org/](chocolatey) package manager for windows and added it to the PATH.
9. Ran `make -f Makefile.simple`
10. Got compiler error about something being uninitialized (`error: missing initializer for member ...`)
11. Realized that there is a mistake in the `Makefile.simple` where `-Wno-maybe-uninitialized` is specified two times for `gcc` toolchain, but is not specified for `mingw` (the flags should be: `-Wno-unknown-warning-option -Wno-maybe-uninitialized -Wno-unused-parameter -Wno-implicit-fallthrough -Wno-missing-field-initializers` for both toolchains)
12. Fixed ti by moving the `-Wno-maybe-uninitialized` to the `mingw` flags
13. Ran `make` again. Got a warning from `mingw.thread.h` starting with `This version of MinGW seems to include a win32 port of...`, but the game compiled successfully. (I also had to modify `C:/ProgramFiles/msys64/mingw64/include/SDL/SDL_gfxPrimitives.h` to replace `<SDL.h>` with `<SDL/SDL.h>`)
14. Started the game. It crashed after one second. Game window appears and closes immediately. Nothing is displayed. It crashes with access violation error. I guess something is not compiled properly.
15. Tried compiling with `-g` option. It took a lot more time to compile. Created a new binary. Started the binary, but it also crashed immediately.
16. Tried to start the binary from WinDbg to figure out why it does not run. WinDbg is unable to load symbols file, because `-g` options works only with `gdb` and has no effects for native windows debuggers.
17. Tried to install `gdb` for windows. Ran the following command:
17.1. `pacman -S mingw-w64-x86_64-gdb`
17.2. Command completed successfully
18. Started the game from the `gdb`. However, it did not crash. Game started, but literally no text appears on screen. Player can move, but HUD text is not displayed.
19. So, if `gdb` is attached, the game does not crash upon startup. So, I tried to set gdb for the default Windows debugger (added a key to the registry editor) and started the game again without `gdb` attached.
20. When game crashed, before closing it, attached a new `gdb` instance to it to analyze the problem.
21. Access violation is caused by a thread which executed instruction at address `0x44c651`. Tried to display the current function using gdb command `list *0x44c651`
22. Gdb crashed with the following error:
```
(gdb) list *0x44c651
0x44c651 is in hr::sdltogl(SDL_Surface*, hr::glfont_t&, int) (basegraph.cpp:377).
terminate called after throwing an instance of 'srchilite::ParserException'
  what():  error during the parsing of a definition file

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
```

23. It seems that it was a [bug](https://sourceware.org/bugzilla/show_bug.cgi?id=25009) in gdb 8.3.0, which is fixed in gdb 8.3.1. However, there is no gdb 8.3.1 currently available for windows as an installable exe binary.

Note: this is unrelated to the issue in the OP. This is related to compiling HyperRogue on Windows. Any clue why it crashes upon startup?

Here is the compiled exe file [hyperrogue.exe](https://github.com/Hakerh400/bin/blob/master/hyperrogue-compiled.zip?raw=true)

**Comment by Hakerh400** [comment]
Closing this issue. There are no clear steps on how to properly build this on Windows and seems like the repository owner is not interested in the issue. This is a great game, but poorly implemented.
